File: leptos-ui/src/components/tool_call/a2ui/blocks.rs
//! Static A2UI block renderers — card, table, kv, status, progress, alert, markdown.

use crate::components::icons::*;
use crate::components::message_turn::{parse_markdown_segments, ContentSegment};
use leptos::prelude::*;

// ── Helpers ─────────────────────────────────────────────────────────

pub fn str_field(data: &serde_json::Value, key: &str) -> Option<String> {
    data.get(key)
        .and_then(|v| v.as_str())
        .map(|s| s.to_string())
}

/// Like str_field but tries a fallback key too.
fn str_field_or(data: &serde_json::Value, primary: &str, fallback: &str) -> Option<String> {
    str_field(data, primary).or_else(|| str_field(data, fallback))
}

fn f64_field_or(data: &serde_json::Value, primary: &str, fallback: &str) -> Option<f64> {
    data.get(primary)
        .and_then(|v| v.as_f64())
        .or_else(|| data.get(fallback).and_then(|v| v.as_f64()))
}

pub fn cell_to_string(v: &serde_json::Value) -> String {
    match v {
        serde_json::Value::String(s) => s.clone(),
        serde_json::Value::Number(n) => n.to_string(),
        serde_json::Value::Bool(b) => b.to_string(),
        serde_json::Value::Null => String::new(),
        other => serde_json::to_string(other).unwrap_or_default(),
    }
}

// ── Block renderers ─────────────────────────────────────────────────

pub fn render_card(data: serde_json::Value) -> impl IntoView {
    let title = str_field(&data, "title").unwrap_or_default();
    let body = str_field_or(&data, "body", "content").unwrap_or_default();
    let icon = str_field(&data, "icon");

    view! {
        <div class="a2ui-card">
            <div class="a2ui-card-header">
                {icon.map(|i| view! { <span class="a2ui-card-icon">{i}</span> })}
                <span class="a2ui-card-title">{title}</span>
            </div>
            {(!body.is_empty()).then(|| view! {
                <div class="a2ui-card-body">{body}</div>
            })}
        </div>
    }
}

pub fn render_table(data: serde_json::Value) -> impl IntoView {
    let headers: Vec<String> = data
        .get("headers")
        .and_then(|v| v.as_array())
        .map(|arr| {
            arr.iter()
                .map(|v| v.as_str().unwrap_or("").to_string())
                .collect()
        })
        .unwrap_or_default();
    let rows: Vec<Vec<String>> = data
        .get("rows")
        .and_then(|v| v.as_array())
        .map(|arr| {
            arr.iter()
                .map(|row| {
                    row.as_array()
                        .map(|cells| cells.iter().map(|c| cell_to_string(c)).collect())
                        .unwrap_or_default()
                })
                .collect()
        })
        .unwrap_or_default();

    view! {
        <div class="a2ui-table-wrap">
            <table class="a2ui-table">
                {(!headers.is_empty()).then(|| view! {
                    <thead>
                        <tr>
                            {headers.iter().map(|h| view! {
                                <th>{h.clone()}</th>
                            }).collect_view()}
                        </tr>
                    </thead>
                })}
                <tbody>
                    {rows.iter().map(|row| view! {
                        <tr>
                            {row.iter().map(|cell| view! {
                                <td>{cell.clone()}</td>
                            }).collect_view()}
                        </tr>
                    }).collect_view()}
                </tbody>
            </table>
        </div>
    }
}

pub fn render_kv(data: serde_json::Value) -> impl IntoView {
    let pairs: Vec<(String, String)> = data
        .get("pairs")
        .and_then(|v| v.as_array())
        .map(|arr| {
            arr.iter()
                .filter_map(|item| {
                    let k = item.get("key")?.as_str()?.to_string();
                    let v = cell_to_string(item.get("value").unwrap_or(&serde_json::Value::Null));
                    Some((k, v))
                })
                .collect()
        })
        .unwrap_or_default();

    view! {
        <div class="a2ui-kv">
            {pairs.iter().map(|(k, v)| view! {
                <div class="a2ui-kv-row">
                    <span class="a2ui-kv-key">{k.clone()}</span>
                    <span class="a2ui-kv-val">{v.clone()}</span>
                </div>
            }).collect_view()}
        </div>
    }
}

pub fn render_status(data: serde_json::Value) -> impl IntoView {
    let label = str_field(&data, "label").unwrap_or("Status".into());
    let level = str_field(&data, "level").unwrap_or("info".into());
    let detail = str_field_or(&data, "detail", "message");
    let cls = format!("a2ui-status a2ui-status-{}", level);

    view! {
        <div class=cls>
            {level_icon(&level, 14)}
            <span class="a2ui-status-label">{label}</span>
            {detail.map(|d| view! { <span class="a2ui-status-detail">{d}</span> })}
        </div>
    }
}

pub fn render_progress(data: serde_json::Value) -> impl IntoView {
    let label = str_field(&data, "label").unwrap_or_default();
    let pct = f64_field_or(&data, "percent", "percentage")
        .unwrap_or(0.0)
        .clamp(0.0, 100.0);
    let width_style = format!("width: {}%", pct);

    view! {
        <div class="a2ui-progress">
            <div class="a2ui-progress-header">
                <span class="a2ui-progress-label">{label}</span>
                <span class="a2ui-progress-pct">{format!("{:.0}%", pct)}</span>
            </div>
            <div class="a2ui-progress-track">
                <div class="a2ui-progress-fill" style=width_style></div>
            </div>
        </div>
    }
}

pub fn render_alert(data: serde_json::Value) -> impl IntoView {
    let message = str_field(&data, "message").unwrap_or_default();
    let level = str_field(&data, "level").unwrap_or("info".into());
    let cls = format!("a2ui-alert a2ui-alert-{}", level);

    view! {
        <div class=cls>
            {level_icon(&level, 14)}
            <span>{message}</span>
        </div>
    }
}

pub fn render_markdown(data: serde_json::Value) -> impl IntoView {
    let content = str_field(&data, "content").unwrap_or_default();
    let segments = parse_markdown_segments(&content);

    view! {
        <div class="a2ui-markdown">
            {segments.into_iter().map(|seg| match seg {
                ContentSegment::Html(html) => view! {
                    <div inner_html=html></div>
                }.into_any(),
                ContentSegment::FencedCode { language, code } => view! {
                    <crate::components::code_block::CodeBlock language=language code=code />
                }.into_any(),
            }).collect_view()}
        </div>
    }
}

// ── Shared icon helper ──────────────────────────────────────────────

fn level_icon(level: &str, size: u32) -> impl IntoView {
    match level {
        "success" => view! { <IconCheckCircle2 size=size /> }.into_any(),
        "warning" => view! { <IconAlertTriangle size=size /> }.into_any(),
        "error" => view! { <IconXCircle size=size /> }.into_any(),
        _ => view! { <IconInfo size=size /> }.into_any(),
    }
}

File: web-ui/src/tool-call/a2ui/blocks.ts
import { esc, sf, sfOr, md, mdInline } from "./types";
// md()      — block-level markdown (headings, lists, code fences, paragraphs)
// mdInline()— inline markdown (bold, italic, code, links — no wrapping <p>)
// esc()     — plain HTML-escape for identifiers, labels, attribute values
import type { A2UIBlock } from "./types";

export function svgIcon(name: string, size = 14): string {
  const s = `width="${size}" height="${size}" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round"`;
  switch (name) {
    case "check-circle": return `<svg ${s}><path d="M22 11.08V12a10 10 0 1 1-5.93-9.14"/><polyline points="22 4 12 14.01 9 11.01"/></svg>`;
    case "alert-triangle": return `<svg ${s}><path d="M10.29 3.86L1.82 18a2 2 0 0 0 1.71 3h16.94a2 2 0 0 0 1.71-3L13.71 3.86a2 2 0 0 0-3.42 0z"/><line x1="12" y1="9" x2="12" y2="13"/><line x1="12" y1="17" x2="12.01" y2="17"/></svg>`;
    case "x-circle": return `<svg ${s}><circle cx="12" cy="12" r="10"/><line x1="15" y1="9" x2="9" y2="15"/><line x1="9" y1="9" x2="15" y2="15"/></svg>`;
    case "info": return `<svg ${s}><circle cx="12" cy="12" r="10"/><line x1="12" y1="16" x2="12" y2="12"/><line x1="12" y1="8" x2="12.01" y2="8"/></svg>`;
    case "external-link": return `<svg ${s}><path d="M18 13v6a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2V8a2 2 0 0 1 2-2h6"/><polyline points="15 3 21 3 21 9"/><line x1="10" y1="14" x2="21" y2="3"/></svg>`;
    default: return "";
  }
}

export function levelIcon(level: string, size = 14): string {
  switch (level) {
    case "success": return svgIcon("check-circle", size);
    case "warning": return svgIcon("alert-triangle", size);
    case "error":   return svgIcon("x-circle", size);
    default:        return svgIcon("info", size);
  }
}

// ── Core blocks ────────────────────────────────────────────────

function renderCard(data: Record<string, unknown>): string {
  const title = sf(data, "title");
  const icon = sf(data, "icon");
  const body = sfOr(data, "body", "content");
  let h = '<div class="a2ui-card">';
  if (title) {
    h += '<div class="a2ui-card-header">';
    if (icon) h += `<span class="a2ui-card-icon">${esc(icon)}</span>`;
    h += `<span class="a2ui-card-title">${esc(title)}</span></div>`;
  }
  if (body) h += `<div class="a2ui-card-body">${md(body)}</div>`;
  h += "</div>";
  return h;
}

function renderTable(data: Record<string, unknown>): string {
  const headers = data.headers as string[] | undefined;
  const rows = data.rows as string[][] | undefined;
  let h = '<div class="a2ui-table-wrap"><table class="a2ui-table">';
  if (headers?.length) {
    h += "<thead><tr>";
    for (const hdr of headers) h += `<th>${esc(hdr)}</th>`;
    h += "</tr></thead>";
  }
  if (rows?.length) {
    h += "<tbody>";
    for (const row of rows) {
      h += "<tr>";
      for (const cell of row) h += `<td>${mdInline(String(cell))}</td>`;
      h += "</tr>";
    }
    h += "</tbody>";
  }
  h += "</table></div>";
  return h;
}

function renderKv(data: Record<string, unknown>): string {
  const pairs = data.pairs as Array<{ key: string; value: string }> | undefined;
  if (!pairs?.length) return "";
  let h = '<div class="a2ui-kv">';
  for (const p of pairs) {
    h += `<div class="a2ui-kv-row"><span class="a2ui-kv-key">${esc(p.key)}</span><span class="a2ui-kv-val">${mdInline(p.value)}</span></div>`;
  }
  h += "</div>";
  return h;
}

function renderStatus(data: Record<string, unknown>): string {
  const label = sf(data, "label");
  const level = sf(data, "level") || "info";
  const detail = sfOr(data, "detail", "message");
  return `<div class="a2ui-status a2ui-status-${esc(level)}">${levelIcon(level)}<span class="a2ui-status-label">${esc(label)}</span>${detail ? `<span class="a2ui-status-detail">${mdInline(detail)}</span>` : ""}</div>`;
}

function renderProgress(data: Record<string, unknown>): string {
  const label = sf(data, "label");
  const pct = Math.max(0, Math.min(100, Number(data.percent ?? data.percentage ?? 0)));
  return `<div class="a2ui-progress"><div class="a2ui-progress-header"><span class="a2ui-progress-label">${esc(label)}</span><span class="a2ui-progress-pct">${pct}%</span></div><div class="a2ui-progress-track"><div class="a2ui-progress-fill" style="width:${pct}%"></div></div></div>`;
}

function renderAlert(data: Record<string, unknown>): string {
  // `message` is canonical; tolerate common aliases so malformed model output
  // does not leave an otherwise valid alert visibly blank.
  const msg = sfOr(data, "message", "content", "body", "text");
  const level = sf(data, "level") || "info";
  return `<div class="a2ui-alert a2ui-alert-${esc(level)}">${levelIcon(level, 16)}<div>${md(msg)}</div></div>`;
}

function renderMarkdown(data: Record<string, unknown>): string {
  const content = sfOr(data, "content", "text");
  return `<div class="a2ui-markdown">${md(content)}</div>`;
}

// Map block type -> renderer
const CORE_RENDERERS: Record<string, (data: Record<string, unknown>) => string> = {
  card: renderCard,
  table: renderTable,
  kv: renderKv,
  status: renderStatus,
  progress: renderProgress,
  alert: renderAlert,
  markdown: renderMarkdown,
};

export { CORE_RENDERERS };
